Update to MBedTLS 3.2.1

joerchan opened the pull request:
Notes:
* Pull requests cannot be accepted until the PR follows the [contributing guidelines](../CONTRIBUTING.md). In particular, each commit must have at least one `Signed-off-by:` line from the committer to certify that the contribution is made under the terms of the [Developer Certificate of Origin](../dco.txt).
* This is just a template, so feel free to use/remove the unnecessary things
## Description
A few sentences describing the overall goals of the pull request's commits.


## Status
**READY/IN DEVELOPMENT/HOLD**

## Requires Backporting
When there is a bug fix, it should be backported to all maintained and supported branches.
Changes do not have to be backported if:
- This PR is a new feature\enhancement
- This PR contains changes in the API. If this is true, and there is a need for the fix to be backported, the fix should be handled differently in the legacy branch

Yes | NO  
Which branch?

## Migrations
If there is any API change, what's the incentive and logic for it.

YES | NO

## Additional comments
Any additional information that could be of interest

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Changelog updated
- [ ] Backported


## Steps to test or reproduce
Outline the steps to test or reproduce the PR here.

TTornblom committed:
BUILD: Update IAR support in CMakeLists.txt

Applied the same change as in mbed-crypto for using this as a sub
project with the IAR toolchain.

Signed-off-by: TTornblom <thomas.tornblom@iar.com>

Patch from TF-M mbedcrypto, applied using:
git am \
modules/tee/tf-m/trusted-firmware-m/lib/ext/mbedcrypto/\
0001-BUILD-Update-IAR-support-in-CMakeLists.txt.patch

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

tamasban committed:
Enable crypto code sharing between independent binaries

Signed-off-by: Tamas Ban <tamas.ban@arm.com>

Patch from TF-M mbedcrypto, applied using:
git am \
modules/tee/tf-m/trusted-firmware-m/lib/ext/mbedcrypto/\
0002-Enable-crypto-code-sharing-between-independent-binar.patch

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

RcColes committed:
Add LMS implementation

Also an LM-OTS implementation as one is required for LMS.

Signed-off-by: Raef Coles <raef.coles@arm.com>

Patch from TF-M mbedcrypto, applied using:
git am \
modules/tee/tf-m/trusted-firmware-m/lib/ext/mbedcrypto/\
0005-Add-LMS-implementation.patch

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

hannestschofenig committed:
[nrf fromtree] CID update to RFC 9146

The DTLS 1.2 CID specification has been published as RFC 9146. This PR updates the implementation to match the RFC content.

Signed-off-by: Hannes Tschofenig <hannes.tschofenig@arm.com>
(cherry picked from commit b93ce795a5e99d6dad6d9e957ee77488f70cb912)
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

frkv committed:
[nrf noup] Undef PSA crypto client in crypto_types

-To remove unnecessary build warnings for define being redefined
 we need to undefine the MBEDTLS_PSA_CRYPTO_CLIENT in the
 crypto_types header file

Signed-off-by: Frank Audun Kvamtrø <frank.kvamtro@nordicsemi.no>
(cherry picked from commit fa39e42f8baa51650584f61924a63889db9eadb4)
(cherry picked from commit 44ea1f53a6240c3cc029c84fca2bc28c2c79d732)
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

Vge0rge committed:
[nrf noup] Remove redefinition ECC_MAX_CURVE_BITS

-This removes the redefinition of the define
 PSA_VENDOR_ECC_MAX_CURVE_BITS which we will set
 in our configuration file.

 Ref: NCSDK-12898

Signed-off-by: Georgios Vasilakis <georgios.vasilakis@nordicsemi.no>
(cherry picked from commit 8373f2e8c3e633abdeeea4846c8f309096576cca)
(cherry picked from commit cd65c48114b41fbdfe989179f718171d9456aa3f)
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

Vge0rge committed:
[nrf noup] Check if ECP_MAX_BITS is set in ecp.h

-The ECP_ALT file sets this variable so we
  will have a multiple definition if we enable
  it in ecp.h.

Signed-off-by: Georgios Vasilakis <georgios.vasilakis@nordicsemi.no>
(cherry picked from commit 8146631993760b0a4c0d67d1ad268a06b0bbde26)
(cherry picked from commit 3aaff4c03fa6e19d0abd28e72af22bb17a1a7d01)
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

Vge0rge committed:
[nrf noup] Check if GCM_C is defined in gcm.h

-This checks if GCM_C is enabled in gcm.h
 before including the functions. This was
 causing build issues when the GCM is disabled
 but GCM_ALT is enabled.

Signed-off-by: Georgios Vasilakis <georgios.vasilakis@nordicsemi.no>
(cherry picked from commit 819f6fe17080428effedaff98e4391886e260425)
(cherry picked from commit 803d4802a145fa3e2bc0c45b37948f7d42e6f97c)
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

Vge0rge committed:
[nrf noup] Add driver_wrapper call for derivation

-This adds a psa_driver_wrapper call for the PSA
 psa_key_derivation_key_agreement API.
 Without this patch the psa_key_derivation APIs
 will never reach the PSA drivers (Oberon, CC3XX)
 and they will only support the software
 implementation. This patch is a noupp because
 we expect the Mbed TLS project to add this
 change in a later release.

Ref: NCSDK-13564

Signed-off-by: Georgios Vasilakis <georgios.vasilakis@nordicsemi.no>
(cherry picked from commit 7f54e37b9f061ce692fdb994ac2a4111eb413c21)
(cherry picked from commit 99fe79b9d761756fdbe0cc37e30327daaa990b71)
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

plskeggs committed:
[nrf noup] Enable use of SNI without x509.

-Enable use of SNI without x509 by testing for
 MBEDTLS_SSL_SERVER_NAME_INDICATION

ref: NCSDK-15193

Signed-off-by: Pete Skeggs <peter.skeggs@nordicsemi.no>
Signed-off-by: Frank Audun Kvamtrø <frank.kvamtro@nordicsemi.no>
(cherry picked from commit ad920c56d83d7b51f8c4ce0db474105b92b85409)
(cherry picked from commit 1ee0f9432129cb481e01af87a2170615241c31ea)
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

plskeggs committed:
[nrf noup] Enable auto-generation of additional TLS/DTLS docs

-Enable more TLS/DTLS types being auto-generated in documentation
 in Mbed TLS.

Note that these are not in use in nRF Connect SDK documentation
generation at the moment, this commit currently has no effect

ref: NCSDK-15193

This one conflicted because PREDEFINED was removed in the doxyfile.
Check if this commit can be dropped.
Conflict resolution is to bring back the old defines.

Signed-off-by: Pete Skeggs <peter.skeggs@nordicsemi.no>
Signed-off-by: Frank Audun Kvamtrø <frank.kvamtro@nordicsemi.no>
(cherry picked from commit c82b26a130b6ad1f71d0873cbbd495a1ee81413c)
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

frkv committed:
[nrf toup] pk_wrap: Add missing MD description for ECDSA verify

-Change from PSA_ALG_ECDSA_ANY to a version that provides
 the correct conversion of md_alg => psa_md_alg

ref: NCSDK-14199

Signed-off-by: Frank Audun Kvamtrø <frank.kvamtro@nordicsemi.no>
(cherry picked from commit bce9d9f3ce17ad64b0572091a5c0635ede639885)
(cherry picked from commit f1b778da14eb12932dde64a9ed2b0c6706cdace1)
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

gstrauss committed:
Add accessors to config DN hints for cert request

mbedtls_ssl_conf_dn_hints()
mbedtls_ssl_set_hs_dn_hints()

Signed-off-by: Glenn Strauss <gstrauss@gluelogic.com>

frkv committed:
[nrf noup] Disable reporting ecjpake kkpp extension for PSK

-Disabling this prevents in-field devices from returning errors
 when non ECJPAKE PSK is used for OpenThread devices.

ref: NCSDK-14629

Signed-off-by: Frank Audun Kvamtrø <frank.kvamtro@nordicsemi.no>
(cherry picked from commit ebf98297b79190e47c237f646435f3cc028a7e4d)
(cherry picked from commit c3a7b7adbecb1d32cce8f183c6a0cc027e610d0a)
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

frkv committed:
[nrf noup] Fix PSA MAC for CMAC key types

-The current version is HMAC-oriented and assumes the key type has
 PSA_KEY_USAGE_SIGN_HASH and PSA_KEY_USAGE_VERIFY_HASH, while
 PSA_KEY_USAGE_SIGN_MESSAGE and PSA_KEY_USAGE_VERIFY_MESSAGE
 is more apt for CMAC.

ref: NCSDK-14656

Signed-off-by: Frank Audun Kvamtrø <frank.kvamtro@nordicsemi.no>
(cherry picked from commit a422f6aea8ffc5a6d9fecdd5720d5e72f12ad6b7)
(cherry picked from commit 4086130ae0607c203a7fd951d08539385c0267ca)
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

Vge0rge committed:
[nrf_noup] Add opaque oberon context for hash

-Since the CC3XX driver can call Oberon to do the
 hashing it needs to be aware of the context size
 of the Oberon operation. This adds this size in
 the build.
-Setting the alignment for the oberon type as it is using
 uint64_t as the first tipe and hence will be 8 byte aligned.

ref: NCSDK-13860
ref: NCSDK-13857

Signed-off-by: Georgios Vasilakis <georgios.vasilakis@nordicsemi.no>
Signed-off-by: Frank Audun Kvamtrø <frank.kvamtro@nordicsemi.no>
(cherry picked from commit f11bcb0ceb9ba7fd6ae508e170bfa0750e30106d)
(cherry picked from commit dfa2fe6d800690d0e7d8db3d420e0877a9ed19ec)
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

mpg committed:
Rm tasks-g2.md

That document was always temporary (said so at the top). Now superseded
by https://github.com/orgs/Mbed-TLS/projects/1#column-18338322

Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>

mpg committed:
Update testing.md

Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>

mpg committed:
Update psa-limitations

- misc updates about on-going/recent work
- removal of the section about mixed-PSK: being done in #5762
- clarifications in some places
- some typo fixes

Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>

mpg committed:
Update strategy.md

Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>

mpg committed:
Clarify effect of USE_PSA on TLS 1.3

The previous wording was wrong, there are parts that are affected.

Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>

mpg committed:
Remove stability waiver from USE_PSA

It was initially motivated by the fact that the PSA Crypto APIs
themselves were not stable. In the meantime, PSA Crypto has reached
1.0.0 so this no longer applies.

If we want user to be able to fully benefit from PSA in order to
isolate long-term secrets, they need to be able to use the new APIs with
confidence. There is no reason to think those APIs are any more likely
to change than any of our other APIs, and if they do, we'll follow the
normal process (deprecated in favour of a new variant).

For reference, the APIs in question are:

mbedtls_pk_setup_opaque() // to use PSA-held ECDSA/RSA keys in TLS

mbedtls_ssl_conf_psk_opaque()   // for PSA-held PSKs in TLS
mbedtls_ssl_set_hs_psk_opaque() // for PSA-held PSKs in TLS

mbedtls_cipher_setup_psa() (deprecated in 3.2)
mbedtls_pk_wrap_as_opaque() (documented internal, to be removed in 3.2)

Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>

mpg committed:
Update docs/use-psa-crypto.md

The scope of the option has been expanded, now it makes more sense to
describe it as "everything except ...".

Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>

leorosen committed:
Fix resource leaks

These potential leaks were flagged by the Coverity static analyzer.

Signed-off-by: Leonid Rozenboim <leonid.rozenboim@oracle.com>

leorosen committed:
Created customary ChangeLog.d entry.

Signed-off-by: Leonid Rozenboim <leonid.rozenboim@oracle.com>

mpg committed:
Remove HKDF-Extract/Expand

Being resolved in https://github.com/Mbed-TLS/mbedtls/issues/5784

Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>

mpg committed:
Clarify the TLS 1.3 situation

Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>

mpg committed:
Now compatible with MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER

Also make a few general clarifications/improvements while at it.

Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>

frkv committed:
[nrf noup] Fix unset key bits for nrf_cc3xx v0.9.14

-The runtime library expects key-bits to be set when it is not
 for cipher and ECDSA, this is fixed here. This may be an issue
 either in Mbed TLS or in nrf_cc3xx v0.9.14. Hence setting as a
 noup

ref: NCSDK-13857

Signed-off-by: Frank Audun Kvamtrø <frank.kvamtro@nordicsemi.no>
(cherry picked from commit 0add9081ba3ab67123ffe2590a1acc6a846a2646)
(cherry picked from commit e14218d5dea3b1d43268909608cb740d36ca2550)
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

superna9999 committed:
Introduce mbedtls_ssl_get_ciphersuite_sig_pk_ext_alg() and use it in ssl_pick_cert()

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

paul-elliott-arm committed:
Add pem_free() to other error paths in pk_parse_public_key()

Signed-off-by: Paul Elliott <paul.elliott@arm.com>

paul-elliott-arm committed:
Improve Changelog

Signed-off-by: Paul Elliott <paul.elliott@arm.com>

superna9999 committed:
Add Cipersuite selection negative testing by using invalid algs for server-side opaque key

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

superna9999 committed:
Add negative testing of ciphersuite selection using Opaque algs & usage

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

ronald-cron-arm committed:
tls13: Simplify switch to the inbound handshake keys on server side

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
tls13: Move state changes up to state main handler

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
tls13: Fix buffer overread checks in ssl_tls13_parse_alpn_ext()

Some coding style alignement as well.

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
tls13: Fix certificate extension size write

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
tls13: Rename ssl_tls13_write_hello_retry_request_coordinate

Rename ssl_tls13_write_hello_retry_request_coordinate to
ssl_tls13_prepare_hello_retry_request as it is more
aligned with what the function does.

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

mpg committed:
Update README about USE_PSA_CRYPTO

Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>

mpg committed:
Link to restartable ECC EPIC

Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>

mpg committed:
Fix list of what's common to TLS 1.2 and 1.3

Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>

mpg committed:
Tune description of PSA crypto implementation

Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>

mpg committed:
Add ChangeLog entries for general Use PSA improvements

Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>

mpg committed:
Misc clean-ups in docs/use-psa-crypto.md

Note: limitations of opaque PSKs changed from "TLS 1.2 only" to "none"
since TLS 1.3 does not support PSK at all so far, and it is expected to
support opaque PSKs as soon as it gains PSK support, it will be just a
matter of selecting between psa_key_derivation_input_bytes() and
psa_key_derivation_input_key() - and testing obviously.

Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>

Commit:
Refactor test_suite_ssl tests to enable cache setting
Signed-off-by: Andrzej Kurek <andrzej.kurek@arm.com>

ronald-cron-arm committed:
tests: ssl: Fix coverity deadcode issue

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

Commit:
Add a test with a bad session_id_len that makes cache setting fail
Force a bad session_id_len before handshake wrapup. This should
result in a forced jump to a clean up of a serialized session.
Signed-off-by: Andrzej Kurek <andrzej.kurek@arm.com>

Commit:
Improve the changelog message to contain more details
Signed-off-by: Andrzej Kurek <andrzej.kurek@arm.com>

Commit:
test_suite_ssl: Add missing arguments in endpoint initialization
Signed-off-by: Andrzej Kurek <andrzej.kurek@arm.com>

Commit:
test_suite_ssl: add missing MBEDTLS_SSL_SERVER_C dependency
Signed-off-by: Andrzej Kurek <andrzej.kurek@arm.com>

Commit:
test_suite_ssl: add missing options cleanup
Signed-off-by: Andrzej Kurek <andrzej.kurek@arm.com>

Commit:
test_suite_ssl: add required dependencies for default handshake parameters
Signed-off-by: Andrzej Kurek <andrzej.kurek@arm.com>

Commit:
Disable bad session id length test in TLS 1.3
Signed-off-by: Andrzej Kurek <andrzej.kurek@arm.com>

ronald-cron-arm committed:
Fix typo in documentation

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

yuhaoth committed:
refactor get sig algo from pk

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
Add pss_rsae sig algs into test conf

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
create tls12/tls13 sig alg support check

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
remove pkcs1_* support

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
Remove pkcs1 from certificate_verify

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
remove default sig_hashes

And add pss_rsae_* sig_algs to fix
`Handshake TLS 1.3` test fails, which
is part of `test_suite_ssl`

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
Revert temp test

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
Rename sig_alg cert_key check

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
remove out of scope codes

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
fix various issues

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
fix unused variable issue

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
Add debug message

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
fix ordered sig algs fail for openssl

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
move big function

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
Remove rsa_pss_rsae_* from tls12 sig_algs

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
Revert order of default sig_algs

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
fix tls12 openssl/gnutls server fail

To test version negotiation with tls12 OpenSSL/GnuTLS server, If
`rsa_pss_rsae_*` were sent to server before `rsa_pkcs_*`, server
will return `rsa_pss_rsae_*` as key exchange sig alg. OpenSSL/GnuTLS
can work with this case. mbedTLS will fail due to `rsa_pss_rsae_*`
unsupported.

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
Revert unnecessary space change

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
remove unnecessary check

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
remove rsa_pkcs1_* from tls13 support list

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
Add tls12 algorithms in hybrid mode client hello

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
Remove pkcs1 from certificate verify.


Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
Add debug message

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
Revert rsa_pss_rsae_* support for tls12

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
fix various issues

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
Remove hack fix for server hybrid issue

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
Add tls13 sig alg parameters

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
add debug messages

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
fix handshake_version test fail.

when both tls13 and tls12 are enabled, the
test will fail.

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
add tests for select sig alg

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

ronald-cron-arm committed:
Improve documentation

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

superna9999 committed:
Rename mbedtls_ssl_get_ciphersuite_sig_pk_ext_XXX in mbedtls_ssl_get_ciphersuite_sig_pk_ext_XXX()

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

superna9999 committed:
Update description for negative key_opaque_algs tests

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

superna9999 committed:
Move preferred_hash_for_sig_alg() check after ssl_pick_cert() and check if hash alg is supported with mbedtls_pk_can_do_ext()

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

gstrauss committed:
Test accessors to config DN hints for cert request

Signed-off-by: Glenn Strauss <gstrauss@gluelogic.com>

yuhaoth committed:
refine supported sig alg print

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
fix various issues

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

Commit:
test_suite_ssl: add proper cache cleanup
Signed-off-by: Andrzej Kurek <andrzej.kurek@arm.com>

superna9999 committed:
Add tests with specific hash for handshake_ciphersuite_select positive tests

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

yuhaoth committed:
Add rsa_pkcs1 for cert sig match

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
Add negative tests

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
fix various issues

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

mpg committed:
Merge pull request #5983 from gstrauss/inline-mbedtls_x509_dn_get_next

Inline mbedtls_x509_dn_get_next() in x509.h

yuhaoth committed:
Remove pkcs1 from key cert and sig alg map

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

mprse committed:
ssl_tls13_parse_certificate_verify(): optimize the code

Signed-off-by: Przemek Stekiel <przemyslaw.stekiel@mobica.com>

gabor-mezei-arm committed:
Move switching to handshake transform after sending CCS record

Signed-off-by: Gabor Mezei <gabor.mezei@arm.com>

gabor-mezei-arm committed:
Typo

Signed-off-by: Gabor Mezei <gabor.mezei@arm.com>

gabor-mezei-arm committed:
Revert "Do not encrypt CCS records"

This reverts commit 96ec83138540d7399d9e17a68f304425661ee9e5.

Signed-off-by: Gabor Mezei <gabor.mezei@arm.com>

ronald-cron-arm committed:
Merge pull request #5908 from ronald-cron-arm/tls13-fixes-doc

TLS 1.3: Fixes and add documentation
Validated by the internal CI, no need to wait for the Open CI.

gabor-mezei-arm committed:
Fix name

Signed-off-by: Gabor Mezei <gabor.mezei@arm.com>

gabor-mezei-arm committed:
Do not force TLS 1.3 on client side for TLS 1.3 middlebox compatibility tests

Signed-off-by: Gabor Mezei <gabor.mezei@arm.com>

Commit:
Formatting fixes
Signed-off-by: Andrzej Kurek <andrzej.kurek@arm.com>

Commit:
Fix changelog wording
Signed-off-by: Andrzej Kurek <andrzej.kurek@arm.com>

gilles-peskine-arm committed:
Merge pull request #5905 from gilles-peskine-arm/changelog-improvements-20220609-development

Changelog improvements before the 3.2 release

gilles-peskine-arm committed:
Merge pull request #5861 from wernerlewis/csr_subject_comma

Fix output of commas and other special characters in X509 DN values

yuhaoth committed:
fix format issue

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

yuhaoth committed:
remove supported check from parse sig algs

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

ronald-cron-arm committed:
Merge pull request #5970 from gabor-mezei-arm/5229_Send_dummy_change_cipher_spec_records_from_server

TLS 1.3 server: Send dummy change_cipher_spec records

The internal CI PR-merge job ran successfully thus good to go.

mpg committed:
Merge pull request #5980 from mprse/md_dep_fix

Remove MD dependencies from mbedtls_x509_sig_alg_gets(), ssl_tls13_parse_certificate_verify()

wernerlewis committed:
Fix spelling and formatting consistency

Signed-off-by: Werner Lewis <werner.lewis@arm.com>

yuhaoth committed:
add tests

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

gilles-peskine-arm committed:
More wording improvements

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
For status values, the macro expansions must not change either

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

daverodgman committed:
Merge pull request #934 from gilles-peskine-arm/mpi-0-mod-2

Fix null pointer dereference in mpi_mod_int(0, 2)

daverodgman committed:
Merge pull request #5976 from gilles-peskine-arm/selftest-calloc-pointer-comparison-fix-development

Remove largely useless bit of test log to silence GCC 12

tom-cosgrove-arm committed:
Document that MBEDTLS_SSL_KEEP_PEER_CERTIFICATE is required by MBEDTLS_SSL_PROTO_TLS1_3

Also have check_config.h enforce this. And MBEDTLS_SSL_EXPORT_KEYS has been removed,
so no longer mention it.

Signed-off-by: Tom Cosgrove <tom.cosgrove@arm.com>

paul-elliott-arm committed:
Merge pull request #939 from ronald-cron-arm/tls13-add-missing-overread-check

TLS 1.3: Add missing overread check

Summer-ARM committed:
Add MBEDTLS_CHACHA20_C and MBEDTLS_POLY1305_C

MBEDTLS_CHACHA20_C and MBEDTLS_POLY1305_C are needed when
PSA_WANT_ALG_CHACHA20_POLY1305 is defined.

Signed-off-by: Summer Qin <summer.qin@arm.com>

Patch from TF-M mbedcrypto, applied using:
git am \
modules/tee/tf-m/trusted-firmware-m/lib/ext/mbedcrypto/\
0003-Add-MBEDTLS_CHACHA20_C-and-MBEDTLS_POLY1305_C.patch

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

superna9999 committed:
Reorganize & add more handshake_ciphersuite_select to test all MBEDTLS_KEY_EXCHANGE_XXX cases

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

gilles-peskine-arm committed:
Add section for Mbed TLS 2.25.0+

We hadn't updated the storage specification in a while. There have been no
changes to the storage layout, but the details of the contents of some
fields have changed.

Since this is now a de facto stable format (unchanged between 2.25 and 3.2),
describe it fully, avoiding references to previous versions.

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

tom-cosgrove-arm committed:
Add comments about MBEDTLS_PSA_CRYPTO_C also being required by MBEDTLS_SSL_PROTO_TLS1_3

Signed-off-by: Tom Cosgrove <tom.cosgrove@arm.com>

superna9999 committed:
Fix depends on handshake_ciphersuite_select tests

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

superna9999 committed:
Reorganize Opaque ssl-opt tests, pass key_opaque_algs=, add less wrong negative server testings

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

mpg committed:
Merge pull request #5981 from mprse/hkdf_config_fix

Add comment to config_psa.h about enabling PSA_HKDF/-EXTRACT/-EXPAND

superna9999 committed:
Add dual keys Opaque ssl-opt tests

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

gilles-peskine-arm committed:
Fix wrong type in C snippet

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Repeat the seed file documentation in 2.25.0

This way the 2.25.0 section is now fully self-contained.

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Typos

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Historical update: the layout on stdio changed in Mbed Crypto 1.1.0

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

ronald-cron-arm committed:
Merge pull request #5612 from tom-cosgrove-arm/tls13-config-options

Document that MBEDTLS_SSL_KEEP_PEER_CERTIFICATE is required by MBEDTLS_SSL_PROTO_TLS1_3

Fully validated by the internal CI. No need to wait for the open one.

ronald-cron-arm committed:
Merge pull request #5917 from gilles-peskine-arm/asn1write-0-fix

Improve ASN.1 write tests

gilles-peskine-arm committed:
Add setbuf platform function

Add a platform function mbedtls_setbuf(), defaulting to setbuf().

The intent is to allow disabling stdio buffering when reading or writing
files with sensitive data, because this exposes the sensitive data to a
subsequent memory disclosure vulnerability.

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Call setbuf when reading or writing files: library

After opening a file containing sensitive data, call mbedtls_setbuf() to
disable buffering. This way, we don't expose sensitive data to a memory
disclosure vulnerability in a buffer outside our control.

This commit adds a call to mbedtls_setbuf() after each call to fopen(),
except:
* In ctr_drbg.c, in load_file(), because this is only used for DH parameters
  and they are not confidential data.
* In psa_its_file.c, in psa_its_remove(), because the file is only opened
  to check its existence, we don't read data from it.

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Call setbuf when reading or writing files: programs

After opening a file containing sensitive data, call mbedtls_setbuf() to
disable buffering. This way, we don't expose sensitive data to a memory
disclosure vulnerability in a buffer outside our control.

This commit adds a call to mbedtls_setbuf() after each call to fopen(),
but only in sample programs that were calling mbedtls_platform_zeroize().
Don't bother protecting stdio buffers in programs where application buffers
weren't protected.

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Changelog entry for mbedtls_setbuf()

* Security: we're improving a countermeasure.
* Requirement change: the library will no longer compile on a platform
  without setbuf().

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

superna9999 committed:
Refine mbedtls_ssl_tls12_get_preferred_hash_for_sig_alg() when USE_PSA_CRYPTO is selected

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

superna9999 committed:
Cleanup Order & Title of Opaque TLS tests, fix RSA- test definition

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

gilles-peskine-arm committed:
Fix Doxygen documentation attached to non-existent elements

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

mpg committed:
Merge pull request #5994 from gilles-peskine-arm/storage-format-doc-2.25-development

Update storage format specification for Mbed TLS 2.25.0+

superna9999 committed:
Use different certs for double opaque keys and check certificate issuer CN

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

superna9999 committed:
Remove auth_mode=required and client crt_file/key_file when testing server authentication

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

ronald-cron-arm committed:
Update TLS 1.3 support documentation

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
Add TLS 1.3 change logs

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

yuhaoth committed:
fix various issues

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

mpg committed:
Merge pull request #5962 from gilles-peskine-arm/storage-format-doc-202206

Documentation about storage format compatibility

mpg committed:
Merge pull request #936 from paul-elliott-arm/fix_tls_record_size_check

Fix the wrong variable being used for TLS record size checks

yuhaoth committed:
remove force_version for client

Signed-off-by: Jerry Yu <jerry.h.yu@arm.com>

ronald-cron-arm committed:
Fix and improve logs and documentation

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

gabor-mezei-arm committed:
Initilize variable

Coverity scan detected a problem of an uinitilized variable.

Signed-off-by: Gabor Mezei <gabor.mezei@arm.com>

superna9999 committed:
Depends on MBEDTLS_X509_REMOVE_INFO disable for double Opaque keys test requiring cert infos to determine selected key

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

superna9999 committed:
Fix mbedtls_ssl_get_ciphersuite_sig_alg() by returning MBEDTLS_PK_NONE for MBEDTLS_KEY_EXCHANGE_RSA

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

ronald-cron-arm committed:
Merge pull request #6005 from ronald-cron-arm/tls13-changelogs-doc-update

TLS 1.3: Add missing change logs and update doc for 3.2 release

Validated by the internal CI, ready to be merged.

paul-elliott-arm committed:
Make definition order a bit neater

Signed-off-by: Paul Elliott <paul.elliott@arm.com>

paul-elliott-arm committed:
Fix changelog requirements section.

Setbuf() is currently not guarded by MBEDTLS_FS_IO, nor can it easily be
so. Raised a seperate issue to cover this, and removed the mention of
MBEDTLS_FS_IO from the Changelog.

Signed-off-by: Paul Elliott <paul.elliott@arm.com>

paul-elliott-arm committed:
Fix code formatting

Signed-off-by: Paul Elliott <paul.elliott@arm.com>

paul-elliott-arm committed:
Merge pull request #5620 from gstrauss/dn_hints

Add accessors to config DN hints for cert request

superna9999 committed:
Fix Handshake select ECDH-RSA- test dependencies

Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

paul-elliott-arm committed:
Merge pull request #5972 from wernerlewis/migration_guide_removals

Add guidance for deprecated functions and macros to migration guide

ronald-cron-arm committed:
Merge pull request #5916 from yuhaoth/pr/tls13-refactor-get-sig-alg-from-pk

Refactor signature algorithm chooser

paul-elliott-arm committed:
Merge pull request #952 from gilles-peskine-arm/stdio_buffering-setbuf

Turn off stdio buffering with setbuf()

Commit:
test_suite_ssl: prefer ASSERT_ALLOC over malloc
Fix formatting for option initialization
Signed-off-by: Andrzej Kurek <andrzej.kurek@arm.com>

Commit:
test_suite_ssl: remove unnecessary user data checks
Signed-off-by: Andrzej Kurek <andrzej.kurek@arm.com>

Commit:
test_suite_ssl: zeroize the cache pointer in case if the struct memory gets reused
Signed-off-by: Andrzej Kurek <andrzej.kurek@arm.com>

mpg committed:
Merge pull request #5835 from superna9999/5831-tls-1-2-ciphersuite-selection

Permissions 2a: TLS 1.2 ciphersuite selection

mpg committed:
Merge pull request #5757 from mpg/update-doc-use-psa

Update "use PSA" documentation (inc. strategy)

Commit:
test_suite_ssl: Fix handshake options cleanup
Signed-off-by: Andrzej Kurek <andrzej.kurek@arm.com>

daverodgman committed:
Merge pull request #6011 from gabor-mezei-arm/coverity_22_07_01

Fix uninitialised memory access in test

Commit:
Improve changelog wording
Co-authored-by: Ronald Cron <ronald.cron@arm.com>
Signed-off-by: Andrzej Kurek <andrzej.kurek@arm.com>

paul-elliott-arm committed:
Merge pull request #5765 from leorosen/fix-some-resource-leaks

Fix resource leaks

Commit:
Clarify the need for calling mbedtls_ssl_derive_keys after extension parsing
Use a more straightforward condition to note that session resumption
is happening.
Co-authored-by: Ronald Cron <ronald.cron@arm.com>
Signed-off-by: Andrzej Kurek <andrzej.kurek@arm.com>

Commit:
Changelog rewording
Signed-off-by: Andrzej Kurek <andrzej.kurek@arm.com>

paul-elliott-arm committed:
Merge pull request #5915 from AndrzejKurek/cid-resumption-clash

Fix DTLS 1.2 session resumption

paul-elliott-arm committed:
Add missing Changelog entries

Signed-off-by: Paul Elliott <paul.elliott@arm.com>

paul-elliott-arm committed:
Rename Changelog entries that don't have .txt extension

Signed-off-by: Paul Elliott <paul.elliott@arm.com>

paul-elliott-arm committed:
Bump library and so versions for 3.2.0 release

Signed-off-by: Paul Elliott <paul.elliott@arm.com>

paul-elliott-arm committed:
Add generated files for 3.2.0 release

Signed-off-by: Paul Elliott <paul.elliott@arm.com>

paul-elliott-arm committed:
Update BRANCHES.md for 3.2.0 release

Signed-off-by: Paul Elliott <paul.elliott@arm.com>

paul-elliott-arm committed:
Fix generation of ssl_debug_helpers

File was being generated with tabs rather than spaces which breaks
release builds

Signed-off-by: Paul Elliott <paul.elliott@arm.com>

daverodgman committed:
Merge pull request #6076 from paul-elliott-arm/fix_generate_debug_helpers

Fix generation of ssl_debug_helpers

ronald-cron-arm committed:
Merge remote-tracking branch 'mbedtls-restricted/development-restricted' into mbedtls-3.2.0rc0-pr

paul-elliott-arm committed:
Update Changelog for 3.2.0 release

Signed-off-by: Paul Elliott <paul.elliott@arm.com>

paul-elliott-arm committed:
Fix formatting of generate_ssl_debug_helpers.py

Satisfy pylint formatting errors

Signed-off-by: Paul Elliott <paul.elliott@arm.com>

paul-elliott-arm committed:
Merge pull request #6077 from paul-elliott-arm/fix_generate_debug_helpers_pylint

Fix formatting of generate_ssl_debug_helpers.py

paul-elliott-arm committed:
Fixup test tls13_server_certificate_msg_invalid_vector_len

The parameters for init_handshake_options had changed on the development
branch after this test was created, so fixing up this test in order to
correct build failures after merge.

Signed-off-by: Paul Elliott <paul.elliott@arm.com>

paul-elliott-arm committed:
Merge remote-tracking branch 'origin/development' into mbedtls-3.2.0rc2-pr

Merge missing commits from development into release

paul-elliott-arm committed:
Ensure return for mbedtls_ssl_write_alpn_ext() is checked

Signed-off-by: Paul Elliott <paul.elliott@arm.com>

paul-elliott-arm committed:
Merge pull request #957 from paul-elliott-arm/mbedtls-3.2.0rc0-pr

Mbedtls 3.2.0rc0 PR

paul-elliott-arm committed:
Revert "Add generated files for 3.2.0 release"

This reverts commit cb21f2eab302b6661405c502ee92f8b8a1f41ae4.

Signed-off-by: Paul Elliott <paul.elliott@arm.com>

paul-elliott-arm committed:
Merge pull request #959 from paul-elliott-arm/merge-into-3.2.0

Merge 3.2.0 into development

daverodgman committed:
Revert "Revert "Add generated files for 3.2.0 release""

This reverts commit 7adb8cbc0ea0bda539b4d1e9aaf9334bf965c8ec.

Signed-off-by: Paul Elliott <paul.elliott@arm.com>

daverodgman committed:
Add missing library/psa_crypto_driver_wrappers.c

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>

paul-elliott-arm committed:
Bump version to 3.2.1

Signed-off-by: Paul Elliott <paul.elliott@arm.com>

paul-elliott-arm committed:
Add 3.2.1 Changelog

Signed-off-by: Paul Elliott <paul.elliott@arm.com>

adeaarm committed:
Driver wrapper entry points for CC3XX

Manually hardcode PSA driver entry points for the CC3XX driver
into psa_crypto_driver_wrappers.c (and provide missing entry point
definitions if any). This is a temporary solution until the codegen
framework is available for automatic integration.

Signed-off-by: Summer Qin <summer.qin@arm.com>
Signed-off-by: Salome Thirot <salome.thirot@arm.com>
Signed-off-by: Abbas Bracken Ziad <abbas.brackenziad@arm.com>
Signed-off-by: Georgios Vasilakis <georgios.vasilakis@nordicsemi.no>
Signed-off-by: Antonio de Angelis <antonio.deangelis@arm.com>

Patch from TF-M mbedcrypto, applied using:
git am \
modules/tee/tf-m/trusted-firmware-m/lib/ext/mbedcrypto/\
0004-Driver-wrapper-entry-points-for-CC3XX.patch

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

RcColes committed:
Add TF-M builtin key driver

Signed-off-by: Raef Coles <raef.coles@arm.com>

Patch from TF-M mbedcrypto, applied using:
git am \
modules/tee/tf-m/trusted-firmware-m/lib/ext/mbedcrypto/\
0006-Add-TF-M-builtin-key-driver.patch

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

Patch from TF-M mbedcrypto, applied using:
git am \
modules/tee/tf-m/trusted-firmware-m/lib/ext/mbedcrypto/\
0007-Build-Remove-encoding-width-suffix-from-Arm-bignum-a.patch

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

adeaarm committed:
CC3XX: Manually enforce no-software builtin fallback when CC3XX is available

This wil be enforced by the autogen framework eventually, but for the time
being we need to manually enforce it for the CC3XX driver only to make sure
that multiple drivers can coexist indepedently.

Signed-off-by: Antonio de Angelis <antonio.deangelis@arm.com>

Patch from TF-M mbedcrypto, applied using:
git am \
modules/tee/tf-m/trusted-firmware-m/lib/ext/mbedcrypto/\
0008-CC3XX-Manually-enforce-no-software-builtin-fallback-.patch

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

david-hazi-arm committed:
Build: Remove encoding width suffix from Arm bignum assembly

Signed-off-by: Dávid Házi <david.hazi@arm.com>

Patch from TF-M mbedcrypto, applied using:
git am \
modules/tee/tf-m/trusted-firmware-m/lib/ext/mbedcrypto/\
0007-Build-Remove-encoding-width-suffix-from-Arm-bignum-a.patch

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

adeaarm committed:
Initialise driver wrappers as first step in psa_crypto_init()

This patch amends the order of initialisations performed in psa_crypto_init()
to make sure that the driver wrappers based on the PSA driver API are initialised
before anything else.

Signed-off-by: Antonio de Angelis <antonio.deangelis@arm.com>

Patch from TF-M mbedcrypto, applied using:
git am \
modules/tee/tf-m/trusted-firmware-m/lib/ext/mbedcrypto/\
0009-Initialise-driver-wrappers-as-first-step-in-psa_cryp.patch

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

Vge0rge committed:
[nrf noup] Zeroize the Mbedtls entropy mutex

This makes sure that the content of the mutex
inside the mbedtls_entropy_context is zeroed.
This is a workaround because the CryptoCell
runtime library will generate a fault if
the mutex is not zeroed. This workaround
will be reverted later when NCSDK-17004
is fixed. There is no reason to upstream this
since it is a limitation in our CryptoCell
runtime library and not an upstream limitation.

Ref: NCSDK-8075

Signed-off-by: Georgios Vasilakis <georgios.vasilakis@nordicsemi.no>
(cherry picked from commit 2bf6a1f10c4239f088a0049d101651ec9fb7b29f)
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

mswarowsky committed:
[nrf noup] Constant size for psa_core_key_attributes_t struct

-There is an inconsistency between PSA Crypto API specification in
 Mbed TLS and in the interface exposed by TF-M for key representation
 where an additional type has been added to hold information about owner.
 This functionality is controlled by setting the configuration
 MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER which configures the type
 Mbed TLS internal type mbedtls_svc_key_id_t to a structure type of two
 words and not as a single word compatible with the PSA Crypto API
 type psk_key_id_t.

 This commit adds a reserved word in psa_core_key_attributes_t after
 the instance of mbedtls_svc_key_id_t to ensure that this structure
 is binary compatible with PSA Crypto drivers that are built with
 MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER.

 This is a [noup] commit as this problem for our pre-built PSA
 crypto drivers which is required to be compiled with the configuration
 MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER enabled to ensure support
 with and without TF-M using the same library.

ref: NCSDK-17464

Signed-off-by: Markus Swarowsky <markus.swarowsky@nordicsemi.no>
Signed-off-by: Frank Audun Kvamtrø <frank.kvamtro@nordicsemi.no>
(cherry picked from commit 2bf6a1f10c4239f088a0049d101651ec9fb7b29f)
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

daverodgman committed:
[nrf fromtree] Fix all.sh dependency on DTLS connection ID

Ensure MBEDTLS_SSL_DTLS_CONNECTION_ID_COMPAT is unset where
MBEDTLS_SSL_DTLS_CONNECTION_ID is unset.

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>
(cherry picked from commit b641ead606205cd90606e2fc459f8c498c77153e)
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

daverodgman committed:
[nrf fromtree] Fix test dependency on DTLS connection ID

Ensure MBEDTLS_SSL_DTLS_CONNECTION_ID and MBEDTLS_SSL_DTLS_CONNECTION_ID_COMPAT
are unset when MBEDTLS_SSL_PROTO_DTLS is not set in tls13-only tests.

Signed-off-by: Dave Rodgman <dave.rodgman@arm.com>
(cherry picked from commit a220ddb64fb46b8de1625c666ac4f82931989288)
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

hannestschofenig committed:
[nrf fromtree] Changed order of conditions in check_config.h

Signed-off-by: Hannes Tschofenig <hannes.tschofenig@arm.com>
Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>
(cherry picked from commit d1425df352b96cba7017e6b1c3c4d3b941cbc051)
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

hannestschofenig committed:
[nrf fromtree] Reference to RFC 9146 added

Added deprecated keyword to MBEDTLS_SSL_DTLS_CONNECTION_ID_COMPAT

Signed-off-by: Hannes Tschofenig <hannes.tschofenig@arm.com>
Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>
(cherry picked from commit 95ee877990e8f1c6bbb2b7e7dd64700069d06eb1)
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

hannestschofenig committed:
[nrf fromtree] Added deprecated warning in check_config.h

Warns about the removal of the legacy DTLS Connection ID feature in a future version of Mbed TLS.

Signed-off-by: Hannes Tschofenig <hannes.tschofenig@arm.com>
Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>
(cherry picked from commit e66b8677b14116bd85c42f5b44479d73121f2cfc)
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

hannestschofenig committed:
[nrf fromtree] Added closing SECTION of doxygen markup

Signed-off-by: Hannes Tschofenig <hannes.tschofenig@arm.com>
Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>
(cherry picked from commit 1a6e9ed691779389c54bfdd3416e7a0467e928c5)
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

hannestschofenig committed:
[nrf fromtree] Removed MBEDTLS_SSL_VARIABLE_BUFFER_LENGTH from CID tests in all.sh

Added also extra text.

Signed-off-by: Hannes Tschofenig <hannes.tschofenig@arm.com>
Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>
(cherry picked from commit 393542bf9a4a429ba866fcce7d7016e7260bc50b)
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

hannestschofenig committed:
[nrf fromtree] Removal of redundant DTLS CID test in ssl-opt.sh

Per suggestion from Manuel, I removed this redundant test.

Signed-off-by: Hannes Tschofenig <hannes.tschofenig@arm.com>
Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>
(cherry picked from commit c743d3ff6a7e015839a2142945ebcc7f2b78cb7d)
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

mpg committed:
[nrf fromtree] Fix bug when legacy CID is enabled but not used

When legacy CID is enabled at compile time, but not used at runtime, we
would incorrectly skip the sequence number at the beginning of the AAD.

There was already two "else" branches for writing the sequence number
but none of them was taken in that particular case.

Simplify the structure of the code: with TLS 1.2 (we're already in that
branch), we always write the sequence number, unless we're using
standard CID.

Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>
(cherry picked from commit 8a4b7322d9c43c564cfa21ac8c139d65c137d11d)
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

mpg committed:
[nrf fromtree] Remove redundant tests in ssl-opt.sh

We don't need to have two copies of the test with one of them depending
on legacy/compat CID: we can have just one copy, but make sure we run
ssl-opt.sh both in a build with standard CID and in a build with
legacy/compat - that's the job of all.sh (see next commit).

Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>
(cherry picked from commit 1de53a172067ffb6ce9d8b2fd796de316c3364fd)
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

mpg committed:
[nrf fromtree] Remove redundant component in all.sh

CID is now enabled in the default config (as well as full), so it's
already tested in numerous all.sh components, not need to add one for
that.

We need a component for the legacy/compat option though as it's never
enabled in existing components. So, keep that one, but adjust the name
and fix a typo in a message.

Signed-off-by: Manuel Pégourié-Gonnard <manuel.pegourie-gonnard@arm.com>
(cherry picked from commit 27ba3989d6dbbfcd90f441eec93e7f4788efa8f9)
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "CC3XX: Hardcode entry points for the CC3XX driver"

This reverts commit 7fed49c9b9f983ad6416986661ef637459723bcb.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "Disable export MbedTLSTargets"

This reverts commit 99bad4d3bee12d22b1ad0496a2fe6f35576d3361.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "BUILD: Update IAR support in CMakeLists.txt"

This reverts commit 6db7216182240a19bbe66da57a37b1c5f5380451.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "mbedtls: remove static keyword from certain function pointers"

This reverts commit 3e3e58a92de76069730c63e0d524f40fea948a61.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Merge tag 'v3.2.1'

Mbed TLS 3.2.1

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

Commit:
Merge pull request for mbedtls-3.2.1

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

joerchan committed:
[ncs merge] Merge commit '6e1660500'

Regular upmerge with MBedTLS from zephyr fork to commit: 6e1660500.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
[nrf toup] Fix RSA dependency on PK write when enabling PSA crypto

Fix RSA (MBEDTLS_RSA_C) dependency on PK write (MBEDTLS_PK_WRITE_C)
when enabling PSA Crypto (MBEDTLS_PSA_CRYPTO_C).
This should instead depend on MBedTLS using PSA crypto
(MBEDTLS_USE_PSA_CRYPTO).

Raised as issue to MbedTLS project:
https://github.com/Mbed-TLS/mbedtls/issues/7126

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
[nrf fromlist] library: psa_crypto: Fix compilation error in psa_key_derivation_abort

Fix compilation error in psa_key_derivation_abort when
MBEDTLS_PSA_BUILTIN_ALG_TLS12_PRF is defined, but
MBEDTLS_PSA_BUILTIN_ALG_TLS12_PSK_TO_MS is not defined.

In function 'psa_key_derivation_abort':
error: 'psa_tls12_prf_key_derivation_t'
{aka 'struct psa_tls12_prf_key_derivation_s'} has no member named
'other_secret'

Upstream PR:
https://github.com/Mbed-TLS/mbedtls/pull/7125

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

SebastianBoe commented on library/pk_wrap.c:
```diff
@@ -105,7 +259,85 @@ static int rsa_verify_wrap( void *ctx, mbedtls_md_type_t md_alg,
 }
+#endif
+
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
```
@joerchan : Vestavind do not permit the toup tag unfortunately.

SebastianBoe replied:
https://github.com/nrfconnect/sdk-mbedtls/pull/15/commits/75b92500d9d2bd83066f7633b56c2bd03322265f

joerchan committed:
[nrf noup] library: psa_crypto: Fix validate tag length function

Fix validate tag length function to validate supported algorithm
from non-builtin drivers.

This change mirrors upstream change in commit:
86679c7bd8856a1a89b4576344554123d8a42c24
 psa_validate_tag_length(): use PSA_WANT_ALG_xxx instead MBEDTLS_PSA_B…
…UILTIN_ALG_xxx guards

However because upstream has removed psa_driver_get_tag_len this change
cannot be cherry-picked.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "[nrf noup] Constant size for psa_core_key_attributes_t struct"

This reverts commit c2a20560148691106ebcb45ff8f1e9557e536c56.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "[nrf fromtree] Make KEY_ID_ENCODES_OWNER compatible with USE_PSA_CRYPTO"

This reverts commit eec8b1d64a847dd123e24fe3a071200775e05027.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "[nrf fromtree] Changelog rewording"

This reverts commit 59e9087a7af45cf73ce6abcd7f3934ead94b2c64.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "[nrf fromtree] Clarify the need for calling mbedtls_ssl_derive_keys after extension parsing"

This reverts commit 96c2601fd677928a5a54aba6b04bc9de7ccbafe6.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "[nrf fromtree] Improve changelog wording"

This reverts commit fc032521f9af13a9c7b64c5772a0e8891c4e9790.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "[nrf fromtree] Add a changelog entry for the session resumption + CID bug"

This reverts commit bf604973c81dbc35cf9194d49c0fd787c3c87835.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "[nrf fromtree] Rearrange the session resumption code"

This reverts commit 3de656b140e93abd38522294e59a8a10bf287cc3.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "[nrf noup]: Zeroize the Mbedtls entropy mutex"

This reverts commit 5ff79b7fec7001c7491b706294d57f20073769db.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "[nrf noup] Fix duplicate declaration of ad_len_field"

This reverts commit c9063745122f4ff470c491ce86c737f0fae24ee7.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "[nrf noup] Fix unset key bits for nrf_cc3xx v0.9.14"

This reverts commit 96f77e3f0a053adb5fb95275a079d61cb4a53966.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "[nrf_noup] Add opaque oberon context for hash"

This reverts commit f1e2cda77f715861ec94be4efc0cceaa10661c40.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "[nrf noup] Fix PSA MAC for CMAC key types"

This reverts commit 10b575d0f7e83e5925c54fc6ec8a0918e1051e6d.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "[nrf noup] Disable reporting ecjpake kkpp extension for PSK"

This reverts commit 6d0cf9e04432c344371bdfbdaf5c040af195940b.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "[nrf toup] pk_wrap: Add missing MD description for ECDSA verify"

This reverts commit 5f51bc40f836726aafdbad68d5f07dea8eeb23b9.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "[nrf noup] Fixes for DTLS CID"

This reverts commit 9e79fcc83e45f6e0340344631ab30bae0c0bc9b8.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "[nrf noup] Backport DTLS CID support compilant with RFC-9146"

This reverts commit 7bf070866e1106600bb8576cf93cede011f35c0a.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "[nrf noup] Enable auto-generation of additional TLS/DTLS docs"

This reverts commit 5c02c487e92deb263dbcf12edcd97feb11a479cd.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "[nrf noup] Enable use of SNI without x509."

This reverts commit 67e2c3ef814a77fd8aa04431c2e1c5d1b4e7c482.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "[nrf noup] Add driver_wrapper call for derivation"

This reverts commit d27e0f0f51b5f566e4c78f397c32aca0d094d833.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "[nrf noup] Check if GCM_C is defined in gcm.h"

This reverts commit c9c142c972d6eaada7fab565a9114384166b2321.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "[nrf noup] Check if ECP_MAX_BITS is set in ecp.h"

This reverts commit f5d7236c8644a59dfb40dee3a1e53e621a0af60a.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "[nrf noup] Remove redefinition ECC_MAX_CURVE_BITS"

This reverts commit 006501de924f021764b4dc33e968247dd6294fc8.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Revert "[nrf noup] Undef PSA crypto client in crypto_types"

This reverts commit 81202bf8f1ff2180e76362a919096d17177aa8eb.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

SebastianBoe commented on library/psa_crypto.c:
```diff
@@ -3876,6 +3734,47 @@ psa_status_t psa_aead_decrypt( mbedtls_svc_key_id_t key,
+
+    switch( PSA_ALG_AEAD_WITH_SHORTENED_TAG( alg, 0 ) )
+    {
+#if defined(PSA_WANT_ALG_CCM)
```
We need a commit message explaining why are changing which defines we use (more detailed explanation than makes it build).

And explaining why it's noup and not fromlist.

https://github.com/nrfconnect/sdk-mbedtls/pull/15/commits/1b53560ea805b8abdbfd13dc1cc24432d72e0ba6

SebastianBoe replied:
The same for

"14e0cd499 * [nrf toup] Fix RSA dependency on PK write when enabling PSA crypto"

SebastianBoe commented on library/psa_crypto.c:
```diff
@@ -4463,6 +4363,15 @@ psa_status_t psa_key_derivation_abort( psa_key_derivation_operation_t *operation
             mbedtls_free( operation->ctx.tls12_prf.label );
         }
 
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_TLS12_PSK_TO_MS)
```
Please add a reference to the upstream PR from the fromlist commit.

For tracking and review purposes.

SebastianBoe commented on library/psa_crypto.c:
```diff
@@ -2380,12 +2380,21 @@ static psa_status_t psa_mac_setup( psa_mac_operation_t *operation,
+                    is_sign ? PSA_KEY_USAGE_SIGN_MESSAGE : PSA_KEY_USAGE_VERIFY_MESSAGE,
+                    alg );
+    }
+
```
This is very weird.

I suspect that when this commit was written the psa_get_and_lock_key_slot_with_policy call above either didn't exist, or looked differently.

And that since then we have resolved a conflict badly.

This commit is invoking the exact same function with the exact same arguments AFAICT which is very rare to be desired behaviour.

https://github.com/nrfconnect/sdk-mbedtls/pull/15/commits/94d7e388c5c408ca45843f393bc0ad5de23a2a10

SebastianBoe replied:
I see the hunk below changes the first invocation from MESSAGE to HASH.

Maybe we need to do the same here?

Perhaps we could ask the author what is going on here.

frkv replied:
Yes. This seems a bit off. 

I think in the newer version that the sign/verify message is already tried, so we only need to try with the sign/verify hash afterwards. 

Remove the additional block trying to lock with sign/verify message (the middle one)

joerchan replied:
Revert and cherry-pick is without conflicts, patch looks the same.
created NCSDK-20159 to follow up on the correctness of the patch itself.

SebastianBoe reviewed (changes requested):
changes requested

joerchan committed:
[nrf fromlist] psa_crypto: Fix psa_key_derivation_output_key ECC without builtin keys

Fix psa_key_derivation_output_key not being able to derive ECC keys
without MBEDTLS_BUILTIN ECC key types enabled.
The PSA crypto drivers can generate these keys without requiring the
builtin key types.

Upstream PR:
https://github.com/Mbed-TLS/mbedtls/pull/7192

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

frkv reviewed (approved):
LGTM pending passing long-running tests
